#**LIBRARY**

In [57]:
# Mengimpor library pandas dan numpy
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.impute import KNNImputer

#**DATASET**

In [58]:
# Memuat data file features.csv dan file labels.csv
features = pd.read_csv("/content/features.csv")
labels = pd.read_csv("/content/labels.csv")

#1. **Menampilkan data yang telah diunduh, dan penjelasan semua kolom yang terdapat pada dataset tersebut.**

In [59]:
# Menampilkan tabel features
features

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 1st sem (without evaluations)  \
0                                                0.0   
1                                                0.0   
2                                                0.0   
3                                                0.0   
4                                                0.0   
...                                              ...   
3091                                             0.0   
3092                                             0.0   
3093                                             0.0   
3094                                             0.0   
3095                                             NaN   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   

In [60]:
# Menampilkan tabel labels
labels

Target
0     Graduate
1     Graduate
2     Graduate
3     Graduate
4      Dropout
...        ...
3091  Graduate
3092  Graduate
3093  Graduate
3094   Dropout
3095  Graduate

[3096 rows x 1 columns]

**Penjelasan Kolom Dataset**

Dataset ini terdiri dari 39 kolom yang memberikan gambaran komprehensif tentang setiap mahasiswa. Berikut adalah rinciannya:

* Unnamed: 0 & Student Id: Kolom identifikasi unik untuk setiap baris data dan mahasiswa.

* Marital status: Status pernikahan mahasiswa.

* Application mode: Jalur pendaftaran yang digunakan mahasiswa (misalnya, tes nasional, transfer, dll.).

* Application order: Urutan pilihan mahasiswa saat mendaftar ke program studi.

* Course: Kode program studi yang diambil oleh mahasiswa.

* Daytime/evening attendance: Waktu perkuliahan (pagi/siang atau malam).

* Previous qualification: Kualifikasi pendidikan terakhir sebelum masuk universitas.

* Previous qualification (grade): Nilai dari kualifikasi pendidikan sebelumnya.

* Nacionality: Kebangsaan mahasiswa.

* Mother's/Father's qualification: Kualifikasi pendidikan terakhir ibu dan ayah.

* Mother's/Father's occupation: Pekerjaan ibu dan ayah.

* Admission grade: Nilai yang diperoleh saat seleksi masuk.

* Displaced: Menunjukkan apakah mahasiswa tersebut merupakan pengungsi atau berasal dari daerah terpencil.

* Educational special needs: Menunjukkan apakah mahasiswa memiliki kebutuhan pendidikan khusus.

* Debtor: Menunjukkan apakah mahasiswa memiliki tunggakan utang.

* Tuition fees up to date: Menunjukkan apakah biaya kuliah mahasiswa lancar.

* Gender: Jenis kelamin mahasiswa.

* Scholarship holder: Menunjukkan apakah mahasiswa penerima beasiswa.

* Age at enrollment: Usia mahasiswa saat pertama kali mendaftar.

* International: Menunjukkan apakah mahasiswa tersebut adalah mahasiswa internasional.

* Curricular units 1st/2nd sem (credited): Jumlah SKS yang diakui/ditransfer pada semester 1 dan 2.

* Curricular units 1st/2nd sem (enrolled): Jumlah SKS yang diambil pada semester 1 dan 2.

* Curricular units 1st/2nd sem (evaluations): Jumlah mata kuliah yang dievaluasi pada semester 1 dan 2.

* Curricular units 1st/2nd sem (approved): Jumlah SKS yang berhasil lulus pada semester 1 dan 2.

* Curricular units 1st/2nd sem (grade): Rata-rata nilai pada semester 1 dan 2.

* Curricular units 1st/2nd sem (without evaluations): Jumlah SKS yang diambil tanpa evaluasi.

* Unemployment rate, Inflation rate, GDP: Indikator sosioekonomi pada saat pendaftaran (tingkat pengangguran, inflasi, dan PDB).

* Target: Variabel target yang menunjukkan status akhir mahasiswa (Graduate, Dropout, atau Enrolled).

#2. **Melakukan agregasi antara tabel features dengan tabel labels.**

In [61]:
df = pd.concat([features, labels], axis = 1)
df

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

#3. **Menampilkan tipe data dari setiap kolom yang terdapat dalam dataset.**

In [62]:

df.dtypes

Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor                                            float64
Tuition fees up to date                           float64
Gender                                            float64
Scholarship holder                                float64
Age at enrollment                                 float64
International                                     float64
Curricular units 1st sem (credited)               float64
Curricular units 1st sem (enrolled)               float64
Curricular units 1st sem (evaluations)            float64
Curricular units 1st sem (approved)               float64
Curricular units 1st sem (grade)                  float64
Curricular units 1st sem (without evaluations)    float64
Curricular units 2nd sem (credited)               float64
Curricular units 2nd sem (enrolled)               float64
Curricular units 2nd sem (evaluations)            float64
Curricular units 2nd sem (approved)               float64
Curricular units 2nd sem (grade)                  float64
Curricular units 2nd sem (without evaluations)    float64
Unemployment rate                                 float64
Inflation rate                                    float64
GDP                                               float64
Target                                             object
dtype: object

In [67]:
# Menampilkan informasi data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3096 entries, 0 to 3095
Data columns (total 39 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Unnamed: 0                                      3096 non-null   int64  
 1   Student Id                                      3096 non-null   int64  
 2   Marital status                                  3012 non-null   object 
 3   Application mode                                2924 non-null   float64
 4   Application order                               2988 non-null   float64
 5   Course                                          2944 non-null   float64
 6   Daytime/evening attendance	                     3016 non-null   object 
 7   Previous qualification                          2973 non-null   float64
 8   Previous qualification (grade)                  3011 non-null   float64
 9   Nacionality                              

# 4. **Statistik Deskriptif dan Analisis Singkat**

In [63]:
# Menampilkan statistik deskriptif
df.describe().T

count         mean  \
Unnamed: 0                                      3096.0  1547.500000   
Student Id                                      3096.0  5470.623062   
Application mode                                2924.0    18.785568   
Application order                               2988.0     1.734940   
Course                                          2944.0  8817.929348   
Previous qualification                          2973.0     4.685503   
Previous qualification (grade)                  3011.0   132.647293   
Mother's qualification                          2951.0    19.470010   
Father's qualification                          3000.0    22.408667   
Mother's occupation                             3034.0    11.495715   
Father's occupation                             2952.0    11.597900   
Admission grade                                 2955.0   126.902876   
Displaced                                       2974.0     0.546066   
Educational special needs                       2977.0     0.010749   
Debtor                                          2980.0     0.117450   
Tuition fees up to date                         2940.0     0.880612   
Gender                                          2894.0     0.350726   
Scholarship holder                              2930.0     0.246416   
Age at enrollment                               2892.0    23.310166   
International                                   2995.0     0.025042   
Curricular units 1st sem (credited)             2888.0     0.698407   
Curricular units 1st sem (enrolled)             2953.0     6.251270   
Curricular units 1st sem (evaluations)          3029.0     8.233080   
Curricular units 1st sem (approved)             2912.0     4.732143   
Curricular units 1st sem (grade)                3007.0    10.692416   
Curricular units 1st sem (without evaluations)  3023.0     0.133311   
Curricular units 2nd sem (credited)             3035.0     0.516639   
Curricular units 2nd sem (enrolled)             2997.0     6.208876   
Curricular units 2nd sem (evaluations)          2889.0     8.061613   
Curricular units 2nd sem (approved)             2945.0     4.422411   
Curricular units 2nd sem (grade)                3021.0    10.275115   
Curricular units 2nd sem (without evaluations)  2943.0     0.139314   
Unemployment rate                               2960.0    11.549561   
Inflation rate                                  2929.0     1.215534   
GDP                                             3032.0     0.028278   

                                                        std      min      25%  \
Unnamed: 0                                       893.882543     0.00   773.75   
Student Id                                      2582.483019  1000.00  3288.50   
Application mode                                  17.532674     1.00     1.00   
Application order                                  1.327527     1.00     1.00   
Course                                          2148.407432    33.00  9085.00   
Previous qualification                            10.439709     1.00     1.00   
Previous qualification (grade)                    13.322662    95.00   124.00   
Mother's qualification                            15.636952     1.00     2.00   
Father's qualification                            15.291008     1.00     3.00   
Mother's occupation                               28.173386     0.00     4.00   
Father's occupation                               26.799023     0.00     4.00   
Admission grade                                   14.460283    95.00   118.00   
Displaced                                          0.497957     0.00     0.00   
Educational special needs                          0.103136     0.00     0.00   
Debtor                                             0.322009     0.00     0.00   
Tuition fees up to date                            0.324299     0.00     1.00   
Gender                                             0.477280     0.00     0.00   
Scholarship holder                

**Analisis Singkat:**

1. Profil Mahasiswa: Rata-rata usia mahasiswa saat mendaftar adalah sekitar 23 tahun, dengan rentang usia yang cukup luas dari 17 hingga 70 tahun. Ini menunjukkan bahwa mayoritas adalah mahasiswa usia normal, tetapi ada juga sebagian kecil mahasiswa dewasa.

2. Kinerja Akademik:

  * Rata-rata nilai masuk (Admission grade) adalah sekitar 126.9 dari skala 190.

  * Mahasiswa rata-rata mengambil 6 SKS (Curricular units enrolled) setiap semester. Namun, jumlah SKS yang lulus (Curricular units approved) rata-rata hanya 4-5 SKS. Ini mengindikasikan adanya tantangan akademik yang dihadapi sebagian mahasiswa.

  * Rata-rata nilai per semester (grade) berada di angka 10-10.7, yang menunjukkan kinerja yang cukup moderat.

3. Latar Belakang Mahasiswa:

  * Sekitar 24.6% mahasiswa adalah penerima beasiswa (Scholarship holder mean = 0.246).

  * Sekitar 11.7% mahasiswa tercatat memiliki tunggakan (Debtor mean = 0.117), yang berpotensi menjadi faktor risiko dropout.

  * Hampir 88% mahasiswa membayar uang kuliah tepat waktu (Tuition fees up to date mean = 0.88), yang merupakan indikator positif.

4. Kualitas Data: Berdasarkan kolom count, hampir semua kolom memiliki nilai yang kurang dari 3096 (jumlah total baris), yang menunjukkan adanya data yang hilang (missing values). Hal ini memvalidasi pentingnya langkah processing data (imputasi) yang telah kamu lakukan menggunakan KNNImputer untuk memastikan dataset siap digunakan untuk pemodelan.

# 5. **SLICING**

* Menampilkan 20 baris awal dari dataset

In [64]:
df.head(20)

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

* Menampilkan 20 baris terakhir

In [65]:
df.tail(20)

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

* Menampilkan index 80-100

In [66]:

df.iloc[80:100]

Unnamed: 0  Student Id Marital status  Application mode  \
80          80        1238         single              17.0   
81          81        4121         single              39.0   
82          82        2890         single               1.0   
83          83        4358         single               1.0   
84          84        4896         single               NaN   
85          85        4735         single              18.0   
86          86        1047         single               1.0   
87          87        7430         single              39.0   
88          88        8527         single               1.0   
89          89        5420         single              43.0   
90          90        5625         single              44.0   
91          91        8990            NaN               1.0   
92          92        9143        married              44.0   
93          93        1944        married               7.0   
94          94        3666         single               1.0   
95          95        6960         single              17.0   
96          96        7620         single               1.0   
97          97        6083         single              42.0   
98          98        5525         single              17.0   
99          99        6029         single              17.0   

    Application order  Course Daytime/evening attendance\t  \
80                1.0  9670.0                          day   
81                1.0  9003.0                          day   
82                1.0   171.0                          day   
83                1.0  9147.0                          day   
84                1.0  9991.0                      evening   
85                1.0  9500.0                          day   
86                1.0  9773.0                          day   
87                2.0  9556.0                          day   
88                2.0  9500.0                          day   
89                1.0     NaN                          day   
90                1.0  9003.0                          day   
91                1.0  9500.0                          day   
92                1.0  9085.0                          day   
93                1.0  9070.0                          day   
94                3.0  9070.0                          day   
95                3.0  9119.0                          day   
96                1.0  9500.0                          day   
97                1.0  9500.0                          day   
98                5.0  9238.0                          day   
99                1.0  9773.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                     1.0                           155.0  Portuguese  ...   
81                     1.0                             NaN  Portuguese  ...   
82                     1.0                           159.0         NaN  ...   
83                     1.0                           150.0   Ukrainian  ...   
84                     1.0                           140.0  Portuguese  ...   
85                     1.0                           144.0  Portuguese  ...   
86                     1.0                           117.0  Portuguese  ...   
87                    19.0                           133.1  Portuguese  ...   
88                     1.0                           145.0  Portuguese  ...   
89                     1.0                           121.0  Portuguese  ...   
90                    39.0                           140.0  Portuguese  ...   
91                     1.0                           141.0  Portuguese  ...   
92                    39.0                           160.0  Portuguese  ...   
93                    43.0                           140.0  Portuguese  ...   
94                     1.0                           125.0  Portuguese  ...   
95                     1.0                           130.0  Portuguese  ...   
96                     1.0        

#6. **Menangani Missing Values Encoding & Imputing**

In [68]:
# Menampilkan missing value sebelum processing
df.isnull().sum()

Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date                           156
Gender                                            202
Scholarship holder                                166
Age at enrollment                                 204
International                                     101
Curricular units 1st sem (credited)               208
Curricular units 1st sem (enrolled)               143
Curricular units 1st sem (evaluations)             67
Curricular units 1st sem (approved)               184
Curricular units 1st sem (grade)                   89
Curricular units 1st sem (without evaluations)     73
Curricular units 2nd sem (credited)                61
Curricular units 2nd sem (enrolled)                99
Curricular units 2nd sem (evaluations)            207
Curricular units 2nd sem (approved)               151
Curricular units 2nd sem (grade)                   75
Curricular units 2nd sem (without evaluations)    153
Unemployment rate                                 136
Inflation rate                                    167
GDP                                                64
Target                                              0
dtype: int64

**encoding terhadap kolom bertipe object menjadi numerik**

In [73]:
# mengubah data object menjadi data numerik dengan labelEncoder
df_proses = df.copy()
df_encode = df_proses.select_dtypes(include=['object']).columns
for col in df_encode:
  le = LabelEncoder()
  df_proses[col] = le.fit_transform(df_proses[col].astype(str))
# Menampilkan update data
df_proses.head()

Unnamed: 0  Student Id  Marital status  Application mode  \
0           0        6634               5               1.0   
1           1        6459               5               1.0   
2           2        2238               5              42.0   
3           3        2479               5               1.0   
4           4        4260               5              44.0   

   Application order  Course  Daytime/evening attendance\t  \
0                3.0  9773.0                             0   
1                1.0     NaN                             0   
2                1.0  9500.0                             0   
3                1.0     NaN                             0   
4                1.0     NaN                             0   

   Previous qualification  Previous qualification (grade)  Nacionality  ...  \
0                     1.0                             NaN           13  ...   
1                     1.0                           136.0           13  ...   
2                     1.0                           120.0           13  ...   
3                     1.0                           141.0           13  ...   
4                    39.0                           120.0           17  ...   

   Curricular units 2nd sem (credited)  Curricular units 2nd sem (enrolled)  \
0                                  0.0                                  6.0   
1                                  0.0                                  0.0   
2                                  4.0                                  8.0   
3                                  0.0                                  6.0   
4                                  1.0                                  6.0   

   Curricular units 2nd sem (evaluations)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     7.0   
4                                     NaN   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                  6.0                         15.333333   
1                                  0.0                          0.000000   
2                                  7.0                         13.285714   
3                                  6.0                         13.142857   
4                                  3.0                         12.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                             0.0               11.1   
1                                             0.0               12.4   
2                                             0.0               12.7   
3                                             0.0               11.1   
4                                             0.0               12.4   

   Inflation rate   GDP  Target  
0             0.6  2.02       2  
1             0.5  1.79       2  
2             3.7 -1.70       2  
3             0.6  2.02       2  
4             0.5  1.79       0  

[5 rows x 39 columns]

**Imputasi dengan model KNNImputer**

In [74]:
# imputasi dengan KNNImputer
numeric_cols = df_proses.select_dtypes(include=np.number).columns
imputer = KNNImputer(n_neighbors=5)
df_proses[numeric_cols] = imputer.fit_transform(df_proses[numeric_cols])

# Menampilkan update data
df_proses.head()

Unnamed: 0  Student Id  Marital status  Application mode  \
0         0.0      6634.0             5.0               1.0   
1         1.0      6459.0             5.0               1.0   
2         2.0      2238.0             5.0              42.0   
3         3.0      2479.0             5.0               1.0   
4         4.0      4260.0             5.0              44.0   

   Application order  Course  Daytime/evening attendance\t  \
0                3.0  9773.0                           0.0   
1                1.0  9130.2                           0.0   
2                1.0  9500.0                           0.0   
3                1.0  5687.8                           0.0   
4                1.0  9388.2                           0.0   

   Previous qualification  Previous qualification (grade)  Nacionality  ...  \
0                     1.0                           141.0         13.0  ...   
1                     1.0                           136.0         13.0  ...   
2                     1.0                           120.0         13.0  ...   
3                     1.0                           141.0         13.0  ...   
4                    39.0                           120.0         17.0  ...   

   Curricular units 2nd sem (credited)  Curricular units 2nd sem (enrolled)  \
0                                  0.0                                  6.0   
1                                  0.0                                  0.0   
2                                  4.0                                  8.0   
3                                  0.0                                  6.0   
4                                  1.0                                  6.0   

   Curricular units 2nd sem (evaluations)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     7.0   
4                                     9.6   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                  6.0                         15.333333   
1                                  0.0                          0.000000   
2                                  7.0                         13.285714   
3                                  6.0                         13.142857   
4                                  3.0                         12.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                             0.0               11.1   
1                                             0.0               12.4   
2                                             0.0               12.7   
3                                             0.0               11.1   
4                                             0.0               12.4   

   Inflation rate   GDP  Target  
0             0.6  2.02     2.0  
1             0.5  1.79     2.0  
2             3.7 -1.70     2.0  
3             0.6  2.02     2.0  
4             0.5  1.79     0.0  

[5 rows x 39 columns]

In [75]:
# Menampilkan missing value setelah pre processing
df_proses.isnull().sum()

Unnamed: 0                                        0
Student Id                                        0
Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance\t                      0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender                                            0
Scholarship holder                                0
Age at enrollment                                 0
International                                     0
Curricular units 1st sem (credited)               0
Curricular units 1st sem (enrolled)               0
Curricular units 1st sem (evaluations)            0
Curricular units 1st sem (approved)               0
Curricular units 1st sem (grade)                  0
Curricular units 1st sem (without evaluations)    0
Curricular units 2nd sem (credited)               0
Curricular units 2nd sem (enrolled)               0
Curricular units 2nd sem (evaluations)            0
Curricular units 2nd sem (approved)               0
Curricular units 2nd sem (grade)                  0
Curricular units 2nd sem (without evaluations)    0
Unemployment rate                                 0
Inflation rate                                    0
GDP                                               0
Target                                            0
dtype: int64

#7. **Menampilkan dataset yang telah melalui seluruh proses preprocessing (cleaning, imputasi, dan encoding).**

In [76]:
# Menampilkan Final DATAFRAME
df_proses

Unnamed: 0  Student Id  Marital status  Application mode  \
0            0.0      6634.0             5.0               1.0   
1            1.0      6459.0             5.0               1.0   
2            2.0      2238.0             5.0              42.0   
3            3.0      2479.0             5.0               1.0   
4            4.0      4260.0             5.0              44.0   
...          ...         ...             ...               ...   
3091      3091.0      6176.0             5.0              44.0   
3092      3092.0      8064.0             5.0               1.0   
3093      3093.0      2103.0             5.0               1.0   
3094      3094.0      8629.0             5.0              39.0   
3095      3095.0      6614.0             5.0              17.0   

      Application order  Course  Daytime/evening attendance\t  \
0                   3.0  9773.0                           0.0   
1                   1.0  9130.2                           0.0   
2                   1.0  9500.0                           0.0   
3                   1.0  5687.8                           0.0   
4                   1.0  9388.2                           0.0   
...                 ...     ...                           ...   
3091                1.0  9085.0                           0.0   
3092                5.0  9238.0                           0.0   
3093                2.0  9500.0                           0.0   
3094                1.0  9500.0                           0.0   
3095                1.0  9670.0                           0.0   

      Previous qualification  Previous qualification (grade)  Nacionality  \
0                        1.0                           141.0         13.0   
1                        1.0                           136.0         13.0   
2                        1.0                           120.0         13.0   
3                        1.0                           141.0         13.0   
4                       39.0                           120.0         17.0   
...                      ...                             ...          ...   
3091                     1.0                           140.0         13.0   
3092                     1.0                           126.0         13.0   
3093                     1.0                           141.0         13.0   
3094                     4.0                           150.0         13.0   
3095                     1.0                           118.0         13.0   

      ...  Curricular units 2nd sem (credited)  \
0     ...                                  0.0   
1     ...                                  0.0   
2     ...                                  4.0   
3     ...                                  0.0   
4     ...                                  1.0   
...   ...                                  ...   
3091  ...                                  1.0   
3092  ...                                  0.0   
3093  ...                                  0.0   
3094  ...                                  2.0   
3095  ...                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        9.6   
...                                      ...   
3091                         